util/deephash/deephash.go

```diff
@@ -164,26 +165,165 @@ func (h *hasher) hashUint64(i uint64) {
 
 var uint8Type = reflect.TypeOf(byte(0))
 
-func (h *hasher) hashValue(v reflect.Value) {
-	if !v.IsValid() {
-		return
+// typeInfo describes properties of a type.
+type typeInfo struct {
+	rtype       reflect.Type
+	isRecursive bool
+
+	atomicComplete int32         // non-zero when complete
+	completeCh     chan struct{} // closed when complete
+
+	// elemTypeInfo is the element type's typeInfo.
+	// It's set when rtype is of Kind Ptr, Slice, Array, Map.
+	elemTypeInfo *typeInfo
+
+	// keyTypeInfo is the map key type's typeInfo.
+	// It's set when rtype is of Kind Map.
+	keyTypeInfo *typeInfo
+}
+
+var typeInfoMap sync.Map           // map[reflect.Type]*typeInfo
+var typeInfoMapPopulate sync.Mutex // just for adding to typeInfoMap
+
+func getTypeInfo(t reflect.Type) *typeInfo {
+	if f, ok := typeInfoMap.Load(t); ok {
+		return f.(*typeInfo)
 	}
+	typeInfoMapPopulate.Lock()
+	defer typeInfoMapPopulate.Unlock()
+	ti := getTypeInfoLocked(t, map[reflect.Type]*typeInfo{})
+	ti.waitComplete()
+	typeInfoMap.Store(t, ti)
+	return ti
+}
 
-	w := h.bw
+func getTypeInfoLocked(t reflect.Type, incomplete map[reflect.Type]*typeInfo) *typeInfo {
+	if v, ok := typeInfoMap.Load(t); ok {
+		return v.(*typeInfo)
+	}
+	if ti, ok := incomplete[t]; ok {
+		return ti
+	}
+	ti := &typeInfo{
+		rtype:       t,
+		isRecursive: typeIsRecursive(t),
+		completeCh:  make(chan struct{}),
+	}
+	incomplete[t] = ti
 
-	if v.CanInterface() {
-		// Use AppendTo methods, if available and cheap.
-		if v.CanAddr() && v.Type().Implements(appenderToType) {
-			a := v.Addr().Interface().(appenderTo)
-			size := h.scratch[:8]
-			record := a.AppendTo(size)
-			binary.LittleEndian.PutUint64(record, uint64(len(record)-len(size)))
-			w.Write(record)
-			return
+	switch t.Kind() {
+	case reflect.Map:
+		ti.keyTypeInfo = getTypeInfoLocked(t.Key(), incomplete)
+		fallthrough
+	case reflect.Ptr, reflect.Slice, reflect.Array:
+		ti.elemTypeInfo = getTypeInfoLocked(t.Elem(), incomplete)
+	}
+
+	close(ti.completeCh)
+	atomic.StoreInt32(&ti.atomicComplete, 1)
+
+	ti.keyTypeInfo.waitComplete()
```

Comment:
Perhaps move:
```go
ti.keyTypeInfo.waitComplete()
ti.elemTypeInfo.waitComplete()
```
to the `typeInfo.waitComplete` method?

Reply:
Disregard. I think I've come to the conclusion that `waitComplete` is unnecessary given that `incomplete` map approach you're taking.